

The following sections describe how to use **qpipeline** to annotate a VCF file with any VCF database file (e.g., COSMIC, dbSNP, ClinVar, _etc._). 

## Create and index a VCF database file 

The instructions provided here can be used to create any VCF database file.  For simplicity, a partial VCF file from NCBI called _common_all_20161122.vcf_ is provided in _${QPIPELINE_HOME}/test_data/vcf_ directory.  This file contained only small number of entries from chromsome 1 downloaded from ftp://ftp.ncbi.nlm.nih.gov/snp/organisms/human_9606/VCF/common_all_20161122.vcf.gz

**NOTE**: Both the input VCF file and the VCF database file must have the same chromosome format. 
That is, either '1',..., '22', 'X', 'Y', 'M' or 'chr1',..., 'chr22', 'chrX', 'chrY', 'chrM'.  Check to make sure your input VCF file and your VCF database file have the same format before indexing your VCF database file.

The sample input VCF file ( _${QPIPELINE_HOME}/test_data/vcf/sample.vcf_ ) uses 'chr1',...,'chr22', 'chrX', 'chrY','chrM' so we will add the prefix 'chr' to _common_all_20161122.vcf_ before indexing it.

Change into directory _${QPIPELINE_HOME}/test_data/vcf_  

```
cd ${QPIPELINE_HOME}/test_data/vcf
```

To create a tabix database file for _common_all_20161122.vcf_, do the following

```
# set FILE variable to "common_all_20161122.vcf" so we can cut, paste, and use the commands below
# for any databases that set to FILE 
FILE="common_all_20161122.vcf"

# get just the VCF header then add 'chr' to all entries
cat $FILE | grep ^# >  ${FILE}.modified.vcf
cat $FILE | grep -v ^# | awk '{ print "chr"$0 }' >> ${FILE}.modified.vcf


# bgzip the newly created file ( bgzip is part of tabix )
bgzip  ${FILE}.modified.vcf


# index the zipped file using tabix
# an index file with extension .tbi is generated by tabix
# Once done, ${FILE}.modified.gz can be used as the VCF database for qpipeline
tabix -p vcf  ${FILE}.modified.vcf.gz 
```


## Annotate an input VCF file against a VCF database file

Annotate _sample.vcf_ with the _common_all_20161122.vcf.modified.vcf.gz_ database and direct output to _sample.common_all.vcf_
```
qpipeline tabix -m 2020  -i sample.vcf -d common_all_20161122.vcf.modified.vcf.gz -q dbSNP142_COMMON_ALL > sample.dbSNP142_COMMON_ALL.vcf 
```
where ( see ${QPIPELINE_HOME}/qpipeline tabix -m 2020 for full usage info )
* -i input VCF file 
* -d bgziped and tabix indexed VCF database file 
* -q VCF database identifer.  For example, dbSNP142_COMMON_ALL.

In the annotation command above, the input VCF file is _sample.vcf_, the bgzipped and tabix indexed VCF database is _common_all_20161122.vcf.modified.vcf.gz_, the VCF database identifier is _dbSNP142_COMMON_ALL_ and the rediected output file is _sample.dbSNP142_COMMON_ALL.vcf_.

If an entry in the input file is matched with one or more entries in the database file ( based on chromosome, position, reference and variant alleles ) then the annotation added to the VCF INFO column looks like as follows:

```
...;IDENTIFER=n,M,MATCHING_DECRIPTIONS,DATA;...
```
where 

* IDENTIFIER is the VCF database identifer entered by the user and n is either 0 (no matches from the VCF database ) or 1 ( there is at least one matched from the VCF database)
* M is the number of matches from the VCF database.   These matches have the same chromosomes, positions, reference (REF) and alternative  (ALT) alleles.
* MATCHING_DECRIPTIONS are the descriptions of the matching record.
* DATA is the matched entries from the VCF database.

The last two lines in the ouput file shows: (1) position chr1:11008 ( C->T ) is not in the dbSNP142_COMMON_ALL database (that is, dbSNP142_COMMON_ALL=0 ) and position chr1:11012 ( C->G ) is in the dbSNP142_COMMON_ALL database with one matching entry (that is, dbSNP142_COMMON_ALL=1,1 ): 
```
chr1    11008   .     C       T       .       .       RS=575272151;dbSNP142_COMMON_ALL=0,0,.
chr1    11012   rs544419019     C       G       .       .       RS=544419019;dbSNP142_COMMON_ALL=1,1,dbSNP142_COMMON_ALL_VARIANT_MATCHED�chr1�11012�rs544419019�C�G�.�.�RS=544419019;RSPOS=11012;dbSNPBuildID=142;SSR=0;SAO=0;VP=0x050000020005150024000100;GENEINFO=DDX11L1:100287102;WGT=1;VC=SNV;R5;ASP;VLD;G5;KGPhase3;CAF=0.9119,0.08806;COMMON=1�
```
***qpipeline*** also allows matches with mismatches REF or ALT allele by using **-R** argument as showed below:
```
qpipeline tabix -m 2020  -i sample.vcf -d common_all_20161122.vcf.modified.vcf.gz -q dbSNP142_COMMON_ALL -R > sample.dbSNP142_COMMON_ALL.vcf 
```

The second last line now shows position chr1:11008 has one match from the dbSNP142_COMMON_ALL database with a mismatched ALT base.  That is, C->T in the input file but C->G in the database file:
```
chr1    11008   .       C       T       .       .       RS=575272151;dbSNP142_COMMON_ALL=1,1,dbSNP142_COMMON_ALL_ALT_BASE_MISMATCHED�chr1�11008�rs575272151�C�G�.�.�RS=575272151;RSPOS=11008;dbSNPBuildID=142;SSR=0;SAO=0;VP=0x050000020005150024000100;GENEINFO=DDX11L1:100287102;WGT=1;VC=SNV;R5;ASP;VLD;G5;KGPhase3;CAF=0.9119,0.08806;COMMON=1�
chr1    11012   rs544419019     C       G       .       .       RS=544419019;dbSNP142_COMMON_ALL=1,1,dbSNP142_COMMON_ALL_VARIANT_MATCHED�chr1�11012�rs544419019�C�G�.�.�RS=544419019;RSPOS=11012;dbSNPBuildID=142;SSR=0;SAO=0;VP=0x050000020005150024000100;GENEINFO=DDX11L1:100287102;WGT=1;VC=SNV;R5;ASP;VLD;G5;KGPhase3;CAF=0.9119,0.08806;COMMON=1�
```
Once annotated, other quick stats can be generated such as:
```
# number of entries in the input file found in the dbSNP142_COMMON_ALL database 
cat  sample.dbSNP142_COMMON_ALL.vcf | grep  dbSNP142_COMMON_ALL=1 | wc -l 

# number of entries in the input file found in the dbSNP142_COMMON_ALL database with exact matches based on chromosomes, positions, REF and ALT alleles
cat  sample.dbSNP142_COMMON_ALL.vcf | grep  dbSNP142_COMMON_ALL_VARIANT_MATCHED | wc -l 
```
Similarly, if the input file is annotated with multiple databases, for example, with dbSNP142, COSMIC69, COSMIC82, and CLINVAR as VCF database identifers, then  many other statistics can be easily extracted:

```
# number of entries found in both COSMIC69 and CLINVAR:
cat FILE.vcf | grep COSMIC69=1 | grep CLINVAR=1 | wc -l

# number of entries in COSMIC82 that are not in COSMIC69
cat FILE.vcf | grep COSMIC69=0 | grep COSMIC82=1 | wc -l
```
